File: src/help_desk_pipeline/source/source_client.py
"""
Klient API dashboardu klienta (source) / Client dashboard API client (source).
Pobiera tickety do przetworzenia / Fetches tickets for processing.
"""
import time
import logging
from typing import List, Optional

import requests

from models import Ticket, TicketStatus
from config import SourceConfig

logger = logging.getLogger(__name__)


class ClientDashboardClient:
    """
    Klient łączący się z dashboardem klienta.
    Client connecting to the client dashboard.
    """

    def __init__(self, config: SourceConfig):
        self.config = config
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {config.api_key}",
            "Content-Type": "application/json",
            "Accept": "application/json",
        })

    def _request_with_retry(self, method: str, url: str, **kwargs) -> requests.Response:
        """
        Wykonaj request z retry w razie błędu.
        Execute request with retry on failure.
        """
        last_exception = None
        for attempt in range(1, self.config.max_retries + 1):
            try:
                response = self.session.request(
                    method, url,
                    timeout=self.config.timeout,
                    **kwargs
                )
                response.raise_for_status()
                return response
            except requests.RequestException as e:
                last_exception = e
                logger.warning(
                    f"Request failed (attempt {attempt}/{self.config.max_retries}): {e}"
                )
                if attempt < self.config.max_retries:
                    time.sleep(self.config.retry_delay * attempt)

        raise ConnectionError(
            f"Failed after {self.config.max_retries} retries: {last_exception}"
        )

    def fetch_tickets(
        self,
        status_filter: Optional[List[TicketStatus]] = None,
        page: int = 1,
        per_page: Optional[int] = None
    ) -> List[Ticket]:
        """
        Pobierz tickety z dashboardu klienta.
        Fetch tickets from the client dashboard.

        Args:
            status_filter: Lista statusów do pobrania / List of statuses to fetch
            page: Numer strony / Page number
            per_page: Liczba wyników na stronę / Results per page

        Returns:
            Lista obiektów Ticket / List of Ticket objects
        """
        per_page = per_page or self.config.batch_size
        params = {
            "page": page,
            "per_page": per_page,
        }
        if status_filter:
            params["status"] = ",".join(s.value for s in status_filter)

        url = f"{self.config.base_url}/tickets"
        logger.info(f"Fetching tickets from {url} (page={page}, per_page={per_page})")

        response = self._request_with_retry("GET", url, params=params)
        data = response.json()

        tickets_raw = data.get("tickets", data.get("data", []))
        tickets = [Ticket.from_dict(raw) for raw in tickets_raw]
        logger.info(f"Fetched {len(tickets)} tickets from client dashboard")
        return tickets

    def fetch_ticket_by_id(self, ticket_id: str) -> Ticket:
        """
        Pobierz pojedynczy ticket po ID.
        Fetch single ticket by ID.
        """
        url = f"{self.config.base_url}/tickets/{ticket_id}"
        logger.info(f"Fetching ticket {ticket_id}")

        response = self._request_with_retry("GET", url)
        ticket_data = response.json()
        return Ticket.from_dict(ticket_data)

    def mark_as_transferred(self, ticket_id: str, helpcenter_id: str) -> bool:
        """
        Oznacz ticket jako przeniesiony do helpcenter.
        Mark ticket as transferred to helpcenter.
        """
        url = f"{self.config.base_url}/tickets/{ticket_id}/status"
        payload = {
            "status": "transferred",
            "helpcenter_ticket_id": helpcenter_id,
        }
        try:
            self._request_with_retry("PATCH", url, json=payload)
            logger.info(f"Marked ticket {ticket_id} as transferred (helpcenter: {helpcenter_id})")
            return True
        except Exception as e:
            logger.error(f"Failed to mark ticket {ticket_id} as transferred: {e}")
            return False

    def get_total_count(self, status_filter: Optional[List[TicketStatus]] = None) -> int:
        """
        Pobierz całkowitą liczbę ticketów do przetworzenia.
        Get total number of tickets to process.
        """
        params = {"count_only": "true"}
        if status_filter:
            params["status"] = ",".join(s.value for s in status_filter)

        url = f"{self.config.base_url}/tickets"
        response = self._request_with_retry("GET", url, params=params)
        data = response.json()
        return data.get("total", 0)

File: src/source_client.rs
//! Source client - Client Dashboard API

use anyhow::Result;
use reqwest::Client;
use std::time::Duration;
use tracing::{debug, error, info, warn};

use crate::config::SourceConfig;
use crate::errors::PipelineError;
use crate::models::{Ticket, TicketStatus};
use crate::security::RateLimiter;

pub struct ClientDashboardClient {
    client: Client,
    config: SourceConfig,
    rate_limiter: RateLimiter,
}

impl ClientDashboardClient {
    pub fn new(config: SourceConfig) -> Result<Self> {
        let client = Client::builder()
            .timeout(Duration::from_secs(config.timeout))
            .pool_max_idle_per_host(10)
            .tcp_keepalive(Duration::from_secs(60))
            .build()?;

        let rate_limiter = RateLimiter::new(config.rate_limit, 60);

        Ok(Self {
            client,
            config,
            rate_limiter,
        })
    }

    /// Fetch tickets from client dashboard
    pub async fn fetch_tickets(
        &self,
        status_filter: Option<Vec<TicketStatus>>,
        page: usize,
    ) -> Result<Vec<Ticket>> {
        self.rate_limiter.wait().await;

        let url = format!("{}/tickets", self.config.base_url);
        let mut request = self
            .client
            .get(&url)
            .bearer_auth(&self.config.api_key)
            .query(&[("page", &page.to_string()), ("per_page", &self.config.batch_size.to_string())]);

        if let Some(statuses) = &status_filter {
            let status_str: Vec<String> = statuses.iter().map(|s| format!("{:?}", s).to_lowercase()).collect();
            request = request.query(&[("status", &status_str.join(","))]);
        }

        info!("Fetching tickets from {} (page={})", url, page);

        let response = request.send().await?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            error!("Failed to fetch tickets: {} - {}", status, body);
            return Err(PipelineError::HttpError(format!("{}: {}", status, body)).into());
        }

        let data: serde_json::Value = response.json().await?;
        let tickets_raw = data
            .get("tickets")
            .or_else(|| data.get("data"))
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        let tickets: Vec<Ticket> = tickets_raw
            .into_iter()
            .filter_map(|raw| serde_json::from_value(raw).ok())
            .collect();

        info!("Fetched {} tickets from client dashboard", tickets.len());
        Ok(tickets)
    }

    /// Fetch single ticket
    pub async fn fetch_ticket_by_id(&self, ticket_id: &str) -> Result<Ticket> {
        self.rate_limiter.wait().await;

        let url = format!("{}/tickets/{}", self.config.base_url, ticket_id);
        debug!("Fetching ticket {}", ticket_id);

        let response = self
            .client
            .get(&url)
            .bearer_auth(&self.config.api_key)
            .send()
            .await?;

        if response.status() == reqwest::StatusCode::NOT_FOUND {
            return Err(PipelineError::TicketNotFound {
                ticket_id: ticket_id.to_string(),
            }
            .into());
        }

        response.error_for_status_ref()?;
        let ticket: Ticket = response.json().await?;
        Ok(ticket)
    }

    /// Mark ticket as transferred
    pub async fn mark_as_transferred(
        &self,
        ticket_id: &str,
        helpcenter_id: &str,
    ) -> Result<bool> {
        self.rate_limiter.wait().await;

        let url = format!("{}/tickets/{}/status", self.config.base_url, ticket_id);
        let payload = serde_json::json!({
            "status": "transferred",
            "helpcenter_ticket_id": helpcenter_id,
        });

        debug!("Marking ticket {} as transferred (helpcenter: {})", ticket_id, helpcenter_id);

        let response = self
            .client
            .patch(&url)
            .bearer_auth(&self.config.api_key)
            .json(&payload)
            .send()
            .await;

        match response {
            Ok(resp) if resp.status().is_success() => {
                info!("✓ Marked ticket {} as transferred", ticket_id);
                Ok(true)
            }
            Ok(resp) => {
                let status = resp.status();
                warn!("Failed to mark ticket {} as transferred: {}", ticket_id, status);
                Ok(false)
            }
            Err(e) => {
                error!("Failed to mark ticket {} as transferred: {}", ticket_id, e);
                Ok(false)
            }
        }
    }

    /// Get total ticket count
    pub async fn get_total_count(&self, status_filter: Option<Vec<TicketStatus>>) -> Result<usize> {
        self.rate_limiter.wait().await;

        let url = format!("{}/tickets", self.config.base_url);
        let mut request = self
            .client
            .get(&url)
            .bearer_auth(&self.config.api_key)
            .query(&[("count_only", "true")]);

        if let Some(statuses) = &status_filter {
            let status_str: Vec<String> = statuses.iter().map(|s| format!("{:?}", s).to_lowercase()).collect();
            request = request.query(&[("status", &status_str.join(","))]);
        }

        let response = request.send().await?;
        let data: serde_json::Value = response.json().await?;
        Ok(data.get("total").and_then(|v| v.as_u64()).unwrap_or(0) as usize)
    }

    /// Health check
    pub async fn health_check(&self) -> Result<(bool, Duration)> {
        let start = std::time::Instant::now();
        let url = format!("{}/health", self.config.base_url);

        let response = self
            .client
            .get(&url)
            .bearer_auth(&self.config.api_key)
            .timeout(Duration::from_secs(5))
            .send()
            .await;

        let latency = start.elapsed();

        match response {
            Ok(resp) if resp.status().is_success() => Ok((true, latency)),
            _ => Ok((false, latency)),
        }
    }

    /// Fetch all pages with pagination
    pub async fn fetch_all_pages(
        &self,
        status_filter: Option<Vec<TicketStatus>>,
    ) -> Result<Vec<Ticket>> {
        let mut all_tickets = Vec::new();
        let mut page = 1;

        loop {
            let batch = self.fetch_tickets(status_filter.clone(), page).await?;

            if batch.is_empty() {
                break;
            }

            all_tickets.extend(batch.clone());

            if batch.len() < self.config.batch_size {
                break; // Last page
            }
            page += 1;
        }

        info!("Fetched {} total tickets across all pages", all_tickets.len());
        Ok(all_tickets)
    }
}
